### PR TITLE
36 fix streams drain control

### LINKED ISSUE #36: fix streams drain control
- [ ] Use new streams in original IntegrationTest
- [ ] Fix a bug in handling DrainControl in streams -- it results in IntegrationTest timeout.

### PR DESCRIPTION
**Pull request by mkrzemien**
Closes #36

### CODE REVIEW

File: core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala [hunk 1]
```diff
@@ -34,7 +34,9 @@ class IntegrationTest
   "KMQ" should "resend message if not committed" in {
     val bootstrapServer = s"localhost:${testKafkaConfig.kafkaPort}"
     val uid = UUID.randomUUID().toString
-    val kmqConfig = new KmqConfig(bootstrapServer, s"$uid-queue", s"$uid-markers", "kmq_client", "kmq_redelivery", 1000, 1000)
+
+    implicit val kmqConfig: KmqConfig = new KmqConfig(bootstrapServer, s"$uid-queue", s"$uid-markers", "kmq_client", "kmq_redelivery", 1000, 1000)
```
Comment (adamw) [inline core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala:38]: probably doesn't need to be implicit

File: core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala [hunk 2]
```diff
@@ -84,24 +86,26 @@ class IntegrationTest
...
     control.shutdown()
   }
 
-  "KMQ" should "resend message if max redelivery count not exceeded" in {
+  it should "resend message if max redelivery count not exceeded" in {
     val bootstrapServer = s"localhost:${testKafkaConfig.kafkaPort}"
     val uid = UUID.randomUUID().toString
-    val kmqConfig = new KmqConfig(bootstrapServer, s"$uid-queue", s"$uid-markers", "kmq_client", "kmq_redelivery", 1000, 1000)
+
+    implicit val kmqConfig: KmqConfig = new KmqConfig(bootstrapServer, s"$uid-queue", s"$uid-markers", "kmq_client", "kmq_redelivery", 1000, 1000)
```
Comment (adamw) [inline core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala:107]: same here

File: core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala [hunk 3]
```diff
@@ -1,37 +1,36 @@
...
 class RedeliveryTrackerStream(markerConsumerSettings: ConsumerSettings[MarkerKey, MarkerValue],
-                              markersTopic: String, maxPartitions: Int)
+                              markersTopic: String)
                              (implicit system: ActorSystem, ec: ExecutionContext,
                               kafkaClients: KafkaClients, kmqConfig: KmqConfig, clock: Clock) extends StrictLogging {
 
   def run(): DrainingControl[Done] = {
     Consumer.committablePartitionedSource(markerConsumerSettings, Subscriptions.topics(markersTopic))
-      .mapAsyncUnordered(maxPartitions) {
+      .map {
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:27]: can you elaborate a bit what was the problem here?

I think this change is incorrect: we'll only ever handle one topic-parition,  as the stream handling that partition's messages will never end (the future returned by `run` will never be complete). So I'd say the previous version was correct

Comment (mkrzemien) [reply core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:27]: Okay, I'm bringing back the mapAsync version. I still need to sort out the problem with broken DrainControl channel.

File: core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala (hunk 1 above)
Comment (mkrzemien) [reply core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala]: Okay

File: core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala (hunk 2 above)
Comment (mkrzemien) [reply core/src/test/scala/com/softwaremill/kmq/redelivery/IntegrationTest.scala]: Okay

File: core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala [hunk 4]
```diff
@@ -1,37 +1,38 @@
...
         case (topicPartition, source) =>
-
-          val redeliverySink = RedeliverySink(topicPartition.partition)
+          val redeliverySink = RedeliverySink(kafkaClients, kmqConfig)(topicPartition.partition)
           val commitMarkerSink = CommitMarkerSink()
 
-          RunnableGraph
-            .fromGraph(GraphDSL.createGraph(redeliverySink, commitMarkerSink)(combineFutures) {
+          val cancellable = RunnableGraph
+            .fromGraph(GraphDSL.createGraph(redeliverySink, commitMarkerSink)(Keep.left) {
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:35]: don't we need to cancel both sinks?

File: core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala [hunk 5]
```diff
@@ -41,10 +42,64 @@ class RedeliveryTrackerStream(markerConsumerSettings: ConsumerSettings[MarkerKey
...
+
+          Future.successful(cancellable)
       }
+      .viaMat(Flow.fromGraph(new AggregatingToMatFlow()))(ControlWithCancellable.apply)
       .toMat(Sink.ignore)(DrainingControl.apply)
       .run()
   }
+}
+
+class AggregatingToMatFlow[T] extends GraphStageWithMaterializedValue[FlowShape[T, T], Iterable[T]] {
```
Comment (adamw) [inline core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:54]: Interesting solution :) @aludwiko or @rucek maybe you can take a look, is this the simplest way to cancel sub-streams created using `.groupBy` or here with `Consumer.committablePartitionedSource`?

File: core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala (hunk 4 above)
Comment (mkrzemien) [reply core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:35]: What's special in the RedeliverySink is an additional TickSource that is instantiated there. Without the TickSource both sinks appear to be closed correctly and the DrainingControl is able to shut down the stream.

Here I'm exposing a hook to the TickSource as Cancellable materialized view from the whole RedeliverySink.

File: core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala (hunk 5 above)
Comment (mkrzemien) [reply core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:54]: I hope it's interesting in a positive way :) 

It's certainly not the simplest in general. But here we have additional TickSource merged inside one of the nested sinks which doesn't stop when it's source stops.

Comment (adamw) [reply core/src/main/scala/com/softwaremill/kmq/redelivery/streams/RedeliveryTrackerStream.scala:54]: Yes, in a positive way :)

I think the problem of stopping sub-streams is not that uncommon, as it emerges also when using the `.groupBy` which I mentioned. So either it's already solved in some way, or we have a solution which would be a great material for a short blog post :)

**Pull request closed without merging**